Contrast Security Integration (#503)
DataDog/integrations-extras: contrastsecurity/README.md

> ## Contrast Security - Protect ScreenBoard for DataDog

Reviewer: ```suggestion
# Contrast Security Integration

## Overview

The Datadog-Constrant integration allows you to get your Contrast logs into Datadog.

## Setup
```

### Set up Contrast Protect logs collection

> You need to enable logs collection for Datadog Agent in `~/.datadog-agent/datadog.yaml` on other paltforms, please find the location of your config file [here](https://docs.datadoghq.com/agent/guide/agent-configuration-files/?tab=agentv6):

Reviewer 2: ```suggestion
Enable logs collection for Datadog Agent in `/etc/datadog-agent/datadog.yaml` on Linux platforms. On other platforms, refer to the [Agent Configuration Files guide](https://docs.datadoghq.com/agent/guide/agent-configuration-files/?tab=agentv6) for the location of your configuration file:
```

```
logs_enabled: true
```

* Create a new folder `java.d` in the `conf.d/` directory of DataDog configuration directory.

> * Create a new conf.yaml file.

Reviewer 2: ```suggestion
* Create a new `conf.yaml` file.
```

> * Create a new folder `java.d` in the `conf.d/` directory of DataDog configuration directory.

Reviewer: Why under java.d instead of the integration config folder?

Author: changed!

> * Create a new folder `java.d` in the `conf.d/` directory of DataDog configuration directory.

Reviewer: ```suggestion
* Add this configuration block to your `contrastsecurity.d/conf.yaml` file to start collecting your Contrast Logs:
```

Author: is this change mandatory? Its just how our docs present it

* Add a custom log collection configuration group.
```
logs:
  - type: file
    path: /path/to/contrast/security.log
    service: contrast
    source: java
```

> * Restart DataDog Agent.

Reviewer 2: ```suggestion
* [Restart the Datadog Agent](https://docs.datadoghq.com/agent/guide/agent-commands/?tab=agentv6#restart-the-agent).
```

>     source: java

Reviewer: ```suggestion
    source: contrastsecurity
```

> More info: https://docs.datadoghq.com/logs/log_collection/?tab=tailexistingfiles#getting-started-with-the-agent

Reviewer 2: ```suggestion
For more information, consult the [Logs Collection documentation](https://docs.datadoghq.com/logs/log_collection/?tab=tailexistingfiles#getting-started-with-the-agent)
```

### Create a Logs Processing Pipeline

> * Create a new pipeline in DataDog dashboard (Logs -> Configuration - > New Pipeline) 

Reviewer 2: ```suggestion
* Create a new pipeline in Datadog (*Logs* -> *Configuration* - > *New Pipeline*) 
```

> * Expand the new pipeline and click Add Processor

Reviewer 2: ```suggestion
* Expand the new pipeline and click *Add Processor*.
```

> * Add a Grok Parser to the pipeline with the following parsing rule

Reviewer 2: ```suggestion
* Add a Grok parser to the pipeline with the following parsing rule:
```

```
ContrastSecurityLogRule %{data:data}pri=%{data:pri} src=%{ip:src} spt=%{number:spt} request=%{data:request} requestMethod=%{word:requestMethod} app=%{data:app} outcome=%{word:outcome}
```

> ### Import Contrast Dashboard into DataDog

Reviewer 2: ```suggestion
### Import contrast dashboard into Datadog
```

Author: Hi. Contrast Security is the name of the company. So could we keep Contrast capitalized

Reviewer 2: Ah, gotcha. Then let's just change it to `Import Contrast dashboard into Datadog`. Note that `Datadog` is not camel cased.

> Your API key and Application keys can be found in Integrations -> APIs

Reviewer 2: ```suggestion
You can find your API key and application keys in *Integrations* -> *APIs*.
```

```bash

> api_key=YOUR_API_KEY

Reviewer 2: ```suggestion
api_key=<YOUR_API_KEY>
```

> app_key=YOUR_APPLICATION_KEY

Reviewer 2: ```suggestion
app_key=<YOUR_APPLICATION_KEY>
```

curl  -X POST -H "Content-type: application/json" -d @contrast_security_protect.json "https://api.datadoghq.com/api/v1/dashboard?api_key=${api_key}&application_key=${app_key}"
```

> More info: https://docs.datadoghq.com/api/?lang=bash#create-a-dashboard

Reviewer 2: ```suggestion
For more information, refer to the [Datadog API documentation for creating a dashboard](https://docs.datadoghq.com/api/?lang=bash#create-a-dashboard).
```

> curl  -X POST -H "Content-type: application/json" -d @contrast_security_protect.json "https://api.datadoghq.com/api/v1/dashboard?api_key=${api_key}&application_key=${app_key}"

Reviewer: Where is `contrast_security_protect.json`? If you include it in the integration and declare it it on the manifest we can create an in-app tile that will allow users to install the screenboard with the click of a button. Note that to do that the json file needs to follow the same format as the ones you find in `integrations-core` repo.

*IMPORTANT*: The dashboard payload that gets placed into the integration folder MUST come from the /screen API endpoint and not the /dash endpoint.

Author: i added it!

Reviewer: Now we do not installing dashboard step :)

Author: Deleted! And short description is there!

> For more information, refer to the [Datadog API documentation for creating a dashboard](https://docs.datadoghq.com/api/?lang=bash#create-a-dashboard).

Reviewer: ```suggestion
For more information, refer to the [Datadog API documentation for creating a dashboard](https://docs.datadoghq.com/api/?lang=bash#create-a-dashboard).

## Data Collected

### Metrics

The Contrast integration does not include any metrics.

### Events

The Contrast integration does not send any events.

### Service Checks

The Contrast integration does not include any service checks.

```

# Contrast Security Integration

## Overview

> The Datadog-Constrant integration allows you to get your Contrast logs into Datadog.

Reviewer 2: ```suggestion
The Datadog-Constant integration allows you to get your Contrast logs into Datadog.
```

# Contrast Security Integration

## Overview

The Datadog-Constrant integration allows you to get your Contrast logs into Datadog.

## Setup

### Set up Contrast Protect logs collection

Enable logs collection for Datadog Agent in `/etc/datadog-agent/datadog.yaml` on Linux platforms. On other platforms, refer to the [Agent Configuration Files guide](https://docs.datadoghq.com/agent/guide/agent-configuration-files/?tab=agentv6) for the location of your configuration file:
```
logs_enabled: true
```

* Create a new folder `java.d` in the `conf.d/` directory of DataDog configuration directory.
* Create a new `conf.yaml` file.
* Add a custom log collection configuration group.
```
logs:
  - type: file

>     path: /path/to/contrast/security.log

Reviewer 3: Is there any doculmentation on where Contrast Security store its logs?
As it would be great to match the default location or to add a link to their documentation that explains where to find them.

Author: Hi for me the path is /Users/<mycomputersname>/.contrast/contrast.log how would that work for datadog?

Reviewer 3: I assume that most users would have this by default.

But we likely want to add a link below this configuration to https://docs.contrastsecurity.com/installation-setupconfig.html#log which seems to be explaining where the logs are stored.
Do you mind adding this?

Author: added

    service: contrast
    source: java
```

* [Restart the Datadog Agent](https://docs.datadoghq.com/agent/guide/agent-commands/?tab=agentv6#restart-the-agent).

For more information, consult the [Logs Collection documentation](https://docs.datadoghq.com/logs/log_collection/?tab=tailexistingfiles#getting-started-with-the-agent).

### Create a Logs Processing Pipeline

* Create a new pipeline in Datadog (*Logs* -> *Configuration* - > *New Pipeline*) 
* Expand the new pipeline and click *Add Processor*.
* Add a Grok parser to the pipeline with the following parsing rule:
```
ContrastSecurityLogRule %{data:data}pri=%{data:pri} src=%{ip:src} spt=%{number:spt} request=%{data:request} requestMethod=%{word:requestMethod} app=%{data:app} outcome=%{word:outcome}
```

> 

Reviewer 3: Thanks for sharing this.

It seems that we can definitely help to provide a more optimised parsing rule (see [keyvalue examples](https://docs.datadoghq.com/logs/processing/parsing/?tab=matcher#key-value)) that would match the global Datadog naming convention so that attribute name would match the same as other integration (`http.method` instead of requestMethod` for instance).

So what we can do there is to create on Datadog side (still not open source but we are working on it) an integration pipeline that would get installed automatically.
For this to happen, would you mind sharing a couple log samples?

Then you will be able to simply remove this part and the integration pipeline would be automatically installed when sending logs with the `contrastsecurity` source.

Author: 2019-11-11 03:17:58,857 [Contrast Server Activity Updater wire] DEBUG -  << "HTTP/1.1 304 Not Modified[\r][\n]"
2019-11-11 03:17:58,857 [Contrast Server Activity Updater wire] DEBUG -  << "Server: Apache-Coyote/1.1[\r][\n]"
2019-11-11 03:17:58,857 [Contrast Server Activity Updater wire] DEBUG -  << "Pragma: no-cache[\r][\n]"
2019-11-11 03:17:58,857 [Contrast Server Activity Updater wire] DEBUG -  << "Expires: Thu, 01 Jan 1970 00:00:00 GMT[\r][\n]"
2019-11-11 03:17:58,857 [Contrast Server Activity Updater wire] DEBUG -  << "Cache-Control: no-cache, no-store[\r][\n]"
2019-11-11 03:17:58,857 [Contrast Server Activity Updater wire] DEBUG -  << "X-Frame-Options: SAMEORIGIN[\r][\n]"
2019-11-11 03:17:58,857 [Contrast Server Activity Updater wire] DEBUG -  << "X-XSS-Protection: 1;mode=block[\r][\n]"
2019-11-11 03:17:58,857 [Contrast Server Activity Updater wire] DEBUG -  << "X-Content-Type-Options: nosniff[\r][\n]"
2019-11-11 03:17:58,857 [Contrast Server Activity Updater wire] DEBUG -  << "Date: Mon, 11 Nov 2019 08:17:58 GMT[\r][\n]"
2019-11-11 03:17:58,857 [Contrast Server Activity Updater wire] DEBUG -  << "[\r][\n]"

Are these suitable example?

Reviewer 3: Thanks for this.

Just wondering if there are other types because when l look at the rule you suggested there are some `request` and `requestMethod` that I cannot see the in the samples.

Author: No. there is not going to be anything more specific request/requestmethod

Reviewer 3: Therefore may I ask where the rules you suggested initially are coming for? For reference:

```
ContrastSecurityLogRule %{data:data}pri=%{data:pri} src=%{ip:src} spt=%{number:spt} request=%{data:request} requestMethod=%{word:requestMethod} app=%{data:app} outcome=%{word:outcome}
```

I just want to make sure we do not miss any interested log formats there.

Author: A co-worker created this rule. I would be open to editing it as needed to whatever it needs to be.

Reviewer 3: Do you know if you could ask this co-worker which logs were used to build this rule?
As I would expect something like:

```
pri=123 src=127.0.0.1 spt=12 request=https://example.com requestMethod=GET app=myapp outcome=success
```

which seems very different from the samples that you shared. So I'm thinking that we likely want to support both format so having examples for this format as well from your co-worker could be super helpful.

Author: ``[INFO] [talledLocalContainer] 261119 13.03.21,597 {} {} {} INFO  (VulnerabilityProcessorQueryService.java:135) [VULNERABILITIES] Saved vulnerability instance property VulnerabilityInstanceProperty(id=19, vulnerabilityInstanceId=0, hash=1805407380, vulnerabilityHash=1805407380, name=/WebGoat/SqlInjection.lesson.lesson, value={"html":"class\u003d\"assignment-success\"\u003e\u003ci class\u003d\"fa fa-2 fa-check hidden\" aria-hidden\u003d\"true\"\u003e\u003c/i\u003e\u003c/div\u003e\n        \u003cform class\u003d\"attack-form\" accept-charset\u003d\"UNKNOWN\" method\u003d\"POST\" name\u003d\"form\" action\u003d\"/WebGoat/SqlInjection/attack5a\" enctype\u003d\"application/json;charset\u003dUTF-8\"\u003e\n            \u003ctable\u003e\n                \u003ctr\u003e\n                    \u003ctd\u003eAccount Name:\u003c/td\u003e\n               ","startIdx":100,"endIdx":257}, type=EVIDENCE)``` is another example log. But I am willing to change for whatever is advised!

Reviewer 3: Thanks for this, that's super helpful.

When you say that you are willing to change for whatever is advised, are you saying that you can change the default logging format for contrast security logs?

As here it seems that the `value` part of the logs are encoded (see the `\u003c`) blocks and on our end it would be super easy to parse this if they did not had those encoded blocks.
Is that something that can be changed on your end?

Author: oh I am willing to change the rules to whatever is advised. I cannot change the default logs unfortunately.

Reviewer 3: Thanks for the clarification.

I have forwarded those examples to the Log team and they will build an integration pipeline that will be automatically installed when the logs are sent with the source `contrastsecurity`.

So no need to add the part about pipeline creation in the documentation.
